src/ar_gateway_middleware.erl

```diff
@@ -94,31 +95,47 @@ handle_labeled_request(Domain, Label, Req) ->
 			not_found(Req)
 	end.
 
-handle_labeled_request_1(Domain, Label, TXID, Filename, Req) ->
+handle_labeled_request_1(Label, TXID, Filename, Req, Env) ->
 	case get_tx_block_hash(TXID) of
 		{ok, BH} ->
-			handle_labeled_request_2(Domain, Label, TXID, Filename, BH, Req);
+			handle_labeled_request_2(Label, TXID, Filename, BH, Req, Env);
 		not_found ->
 			not_found(Req)
 	end.
 
-handle_labeled_request_2(Domain, Label, TXID, Filename, BH, Req) ->
+handle_labeled_request_2(Label, TXID, Filename, BH, Req, Env) ->
 	case ar_domain:derive_tx_label(TXID, BH) of
-		Label -> serve_tx(Filename, Req);
-		OtherLabel -> redirect_to_labeled_tx(Domain, OtherLabel, TXID, Req)
+		Label ->
+			handle_labeled_request_3(Label, TXID, Filename, Req, Env);
+		OtherLabel ->
+			redirect_to_labeled_tx(OtherLabel, TXID, Req, Env)
+	end.
+
+handle_labeled_request_3(Label, TXID, Filename, Req, Env) ->
+	case cowboy_req:scheme(Req) of
+		<<"https">> -> serve_tx(Filename, Req);
+		<<"http">> -> redirect_to_labeled_tx(Label, TXID, Req, Env)
 	end.
 
-handle_custom_request(Domain, CustomDomain, Req) ->
+handle_custom_request(CustomDomain, Req, Env) ->
 	case resolve_path(cowboy_req:path(Req)) of
 		root -> handle_custom_request_1(CustomDomain, Req);
-		{tx, TXID, _} -> derive_label_and_redirect(Domain, TXID, Req);
+		{tx, TXID, _} -> derive_label_and_redirect(TXID, Req, Env);
+		other -> other_request(Req, Env);
 		invalid -> bad_request(Req)
 	end.
 
+handle_unknown_request(Req, Env) ->
+	case cowboy_req:scheme(Req) of
+		<<"https">> -> not_found(Req);
+		<<"http">> -> other_request(Req, Env)
+	end.
+
 resolve_path(Path) ->
 	case Path of
 		<<"/">> -> root;
-		<<"/", Hash/binary>> -> resolve_path_1(Hash);
+		<<"/", Hash:43/bytes>> -> resolve_path_1(Hash);
```

Comment:
There could be a valid path, like `/whatever/whateverelse/...`, which could happen to be 43 bytes long. Should we instead return `other` if we fail to recognize a TX ID?

Reply:
Good catch! Fixed in 75e950b